### DISCUSSION TITLE
TPMS sensors

### DISCUSSION THREAD

**Discussion by j1mco** [jhansche/ha-teslafi#41] [Ideas]
Please would it be possible to include the tpms sensors for tyre pressures. It's useful for monitoring purposes incase they need topping up time to time:)

**Comment by jhansche** [comment]
I would if I could. Unfortunately TeslaFi's API does not provide these values, despite showing the TPMS readings in their own UI. This would need to be a feature request to TeslaFi.

**Reply by j1mco** [reply]
Thank you. I've messaged teslafi support about this. Seems odd they would like omit this data from the API. Have you seen anything else that is missing?

**Comment by j1mco** [comment]
Hey @jhansche, thanks again for the response. Teslafi have replied back already and claimed that TPMS is in the vehicle data.  Would it be possible for you to implement this? Thanks for the excellent work with this integration, it is excellent and a saviour since Tesla changed to Fleet APIs!

**Reply by jhansche** [reply]
Indeed - it is there now! It was not present when the integration started, or last time I checked for it. This would be a nice addition! Thank you for reaching out to them.

**Reply by jhansche** [reply]
FWIW, it's not reporting accurate values for my vehicle:
```json
{
  "pressure": "psi",
  "tpms_front_left": "0.0",
  "tpms_front_right": "0.0",
  "tpms_rear_left": "0.0",
  "tpms_rear_right": "0.0"
}
```
Despite showing correct values at teslafi.com:
![Screenshot 2024-06-10 at 1 19 37 PM](https://github.com/jhansche/ha-teslafi/assets/99272/77552453-f31c-40fc-aea8-f80cb799f9e4)

Can you confirm?

**Reply by j1mco** [reply]
I'll ping them

**Comment by j1mco** [comment]
Quick response from teslafi, apparently it's "live" data, so it will only populate while the car is driving. Can home assistant handle that?

**Reply by jhansche** [reply]
We only hit `command=lastGood`, so as long as that includes it when actively driving, then yes. It'll just have gaps while you're not driving. If I mark it as `Unavailable` when not actively driving or when it reports 0.0, then that should work, it will just appear choppy and with gaps.

It would look just like the `Speed` graph, but more stable because the tire pressure won't fluctuate as much as speed does 😄 
![Screenshot 2024-06-10 at 7 40 40 PM](https://github.com/jhansche/ha-teslafi/assets/99272/24003780-daae-441a-bc3b-a22bd04a1fa5)

**Reply by jhansche** [reply]
Looking at my own Data Feed on TeslaFi, I see TPMS data would be present while the car is driving, so yes this should work.
![Screenshot 2024-06-10 at 7 55 21 PM](https://github.com/jhansche/ha-teslafi/assets/99272/62ccf4aa-ce18-4791-ab5a-daebbeaa15a4)

Thanks for the suggestion! I'll work on this as soon as I have a chance - hopefully over the next couple weeks.

**Comment by jhansche** [comment]
I just did a test, and even when TeslaFi API was updating with driving data, the `tpms_*` fields were all still `0.0`. I've run into this in the past as well - I see data on the TeslaFi website, but that data isn't making its way into the TeslaFi API response.

I've opened #42, but I'm not convinced that this is going to work. I'll do some more testing to see if it shows up if we remove `command=lastGood`.

@j1mco if you want to try out the PR branch `pr/fix-41/tpms`, and you used HACS to install the integration, and you've enabled HACS experimental features, you can install the branch version using service update.install:
```
service: update.install
target:
  entity_id: update.teslafi_update
data:
  version: pr/fix-41/tpms
```

The 4 entities are in the Diagnostic group, and they are disabled by default so you need to enable them. But I've confirmed in the logs that the actual JSON data we get back from TeslaFi still contains `0.0` for these.

**Reply by j1mco** [reply]
Thanks for testing, i've gone back to Teslafi again... we may have stumbled on a bug their end!

**Reply by j1mco** [reply]
Ok, there should be a fix from teslafi now and it should also show the last reading too

**Reply by jhansche** [reply]
Yeah I noticed the data started filling in for me over the weekend. Looks like it's working. I might still want to mark it unavailable when the car is not driving, because otherwise the graph is going to look broken.

**Comment by j1mco** [comment]
Awesome. I'm thinking the last known on some of these variables make sense
because in theory that's what it should be - or an option to choose last
known or unavailable?

Safe to say - your integration works really well, thank you!

**Reply by jhansche** [reply]
Tire pressure fluctuates as a function of temperature, which fluctuates from the environment as well as from the friction of driving. So last known is fine as a general guide, but not because it's "what it should be" 😬 

The graph will jump from last known to last changed, which could be very drastic, since the tires will have cooled down by the time you start driving again. 

I don't really want to go overboard on option overload, but I'll think on this for a while.

**Comment by jhansche** [comment]
Following up here, this is what my graph looks like when the API is returning data while the car is asleep. It's not bad, and like you said it kinda gives an overview of tire pressure over time. The part I was worried about is how it tends to change dramatically when it does change (i.e. this is when the car starts driving, compared to the long periods of constant values while sleeping).

I could be convinced either way 🤷‍♂️

![Screenshot_20240621-121144](https://github.com/jhansche/ha-teslafi/assets/99272/391d4b79-e7ae-4a55-af67-a105fea7bd2f)

**Comment by jhansche** [comment]
Released in v2.7
